src/moments.jl

```diff
@@ -158,6 +158,39 @@ function mean_and_std(x::RealArray, w::AbstractWeights, dims::Int;
 end
 
 
+"""
+    n_mean_var(x; mean=nothing, corrected=true) -> (n, mean, var)
+
+Return the length, mean, and variance of `x`.
+"""
+function n_mean_var(x; mean=nothing, corrected=true)
+    if isempty(x)
+        # Return the NaN of the type that we would get for a nonempty x
+        n = 0
+        variance = var(x)
+        mean = Statistics.mean(x)
+    elseif mean ≢ nothing
+        n, sse = mapreduce(.+, x) do element
+            return 1, abs2(element - mean)
+        end
+        variance = sse / (n - corrected)
+    else
+        # Use Welford algorithm as seen in (among other places)
+        # Knuth's TAOCP, Vol 2, page 232, 3rd edition.
+        n = 0
+        mean = zero(eltype(x))
+        sse = real(zero(mean))
+        for element in x
+            n += 1
+            new_mean = mean + (element - mean) / n
+            sse += realXcY(element - mean, element - new_mean)
+            mean = new_mean
+        end
+        variance = sse / (n - corrected)
+    end
+    return n, mean, variance
+end
+
```

Comment:
```suggestion
```

Reply:
As mentioned above, this is necessary because not all iterators have a `length` method defined.

Reply:
Ah, I see. In this case, however, I suggest reverting to the current implementation of `sem` on the master branch but
- add support for a precomputed `mean` and
- add a method for `AbstractArray`s that allows to use pairwise summation, existing optimizations etc. as
```julia
sem(x::AbstractArray; mean=nothing) = sqrt(var(x; mean=mean, corrected=true) / length(x))
```
I don't think it's good to add this as a separate function since it is only used by `sem` and factoring out into a separate function does not seem to provide any benefits in the implementation of `sem` (maybe it makes the code more readable though if you use the pattern `sem(x; mean=nothing) = _sem(x, mean)` and dispatch on `mean`). Also don't use `mapreduce(.+, ...)` etc. as commented above.

Reply:
> I don't think it's good to add this as a separate function since it is only used by `sem` and factoring out into a separate function does not seem to provide any benefits in the implementation of `sem` (maybe it makes the code more readable though if you use the pattern `sem(x; mean=nothing) = _sem(x, mean)` and dispatch on `mean`). 

Mostly I factored it out because there are other functions that actually could be improved by taking advantage of this, but I've left those for a future PR. See #753.

> Also don't use `mapreduce(.+, ...)` etc. as commented above.
In general, or for the optimized function? I could rewrite it as a loop but I don't think it'd make any difference.

Reply:
I'm very certain it makes a difference. Broadcasting stresses the compiler quite a bit and usually loops are optimized much more (and pairwise summation won't be performed with `mapreduce` either).

Reply:
> Mostly I factored it out because there are other functions that actually could be improved by taking advantage of this, but I've left those for a future PR. See #753.

It makes the code messy and less readable without any advantage in this PR. If other functions are changed at some point, it might make sense to refactor the code. But there's no benefit right now.

Reply:
> I'm very certain it makes a difference. Broadcasting stresses the compiler quite a bit and usually loops are optimized much more (and pairwise summation won't be performed with `mapreduce` either).

1. What do you mean by "Stresses the compiler?" I don't think I've ever heard that, and it definitely sounds very weird.
2. `mapreduce` is different from broadcasting.
3. Using `reduce` [does actually get rewritten as pairwise summation.](https://discourse.julialang.org/t/reduce-vs-foldl-performance-and-precision/43231/10) Similarly, the docstring for `reduce` mentions:
> The associativity of the reduction is implementation-dependent; if you need a particular associativity, e.g. left-to-right, you should write your own loop or consider using foldl or foldr. 

Cthulhu.jl confirms this, with `mapreduce` eventually being rewritten to:
```
function mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted,
                                  ifirst::Integer, ilast::Integer, blksize::Int)
    if ifirst == ilast
        @inbounds a1 = A[ifirst]
        return mapreduce_first(f, op, a1)
    elseif ilast - ifirst < blksize
        # sequential portion
        @inbounds a1 = A[ifirst]
        @inbounds a2 = A[ifirst+1]
        v = op(f(a1), f(a2))
        @simd for i = ifirst + 2 : ilast
            @inbounds ai = A[i]
            v = op(v, f(ai))
        end
        return v
    else
        # pairwise portion
        imid = ifirst + (ilast - ifirst) >> 1
        v1 = mapreduce_impl(f, op, A, ifirst, imid, blksize)
        v2 = mapreduce_impl(f, op, A, imid+1, ilast, blksize)
        return op(v1, v2)
    end
end
```

Reply:
I think we should use exactly the same implementation as `var` in Statistics (which has one method for general iterables and one method for arrays) for consistency. If something can be improved, let's change it there first and then in StatsBase. (One day we'll merge the two packages and get rid of this annoying duplication.)

Reply:
1. The broadcasting machinery is quite involved, and hence it increases compilation time (compiler has to figure out eg types and shapes).
2. Sure but `.+` is a broadcasting operation.
3. I did not talk about `reduce`.
4. This applies only to Array and Broadcasted inputs - it is not possible to use pairwise summation without cache and additional allocations for general iterables. But the Array case is already handled by the default `var` implementation in an efficient way, and `length` is cheap for it. So it's simpler and efficient to just define `sem(x::AbstractArray; mean=nothing) = sqrt(var(x; mean=mean, corrected=true) / length(x))`.